### PR TITLE
Add react-testing-library

### PR DESCRIPTION
**Pull request by vpicone**
Initial setup for #4233 

#### Changelog

**New**
- [@testing-library/react](https://github.com/testing-library/react-testing-library) –  React DOM testing utilities
- [@testing-library/jest-dom](https://github.com/testing-library/jest-dom#table-of-contents) – jest matchers for inspecting the DOM (e.g. toHaveFocus, toHaveClass, toHaveAttribute)

**Changed**
- Use react testing library for render
- Accordion wrapper tests

**Removed**
- snapshot test

#### Testing / Reviewing
**For any updates to test files, reviewers should raise concerns if they feel a removed or altered test was not adequately replaced.**
- `yarn install && yarn test`

### CODE REVIEW

**Comment by joshblack** [comment]
@vpicone how'd we want to handle running axe btw? Anything we could hook into from RTL?

**Comment by vpicone** [comment]
@joshblack jest-axe officially works with RTL: https://github.com/nickcolley/jest-axe#testing-react-with-react-testing-library instead of using axe-core directly. Do you want me to swap those out in this PR?

File: packages/react/src/components/Accordion/__tests__/Accordion-test.js [hunk 1]
```diff
@@ -5,30 +5,20 @@
  * LICENSE file in the root directory of this source tree.
  */
 
-import { render, cleanup } from '@carbon/test-utils/react';
-import { mount } from 'enzyme';
+// import { render, cleanup } from '@carbon/test-utils/react';
```
Comment (emyarod) [inline packages/react/src/components/Accordion/__tests__/Accordion-test.js]: this isn't needed right?

```suggestion
```

**Comment by joshblack** [comment]
@vpicone I thought we talked earlier that this was kind of annoying to have to explicitly type out each time, especially after each interaction, right? Would be great if `render` applied this automatically and if we had an `act` helper that would run axe afterwards.

**Comment by joshblack** [comment]
Lol totally wrong button, my bad!

**Comment by vpicone** [comment]
@joshblack we could disable RTL's automatic cleanup, then use jest's afterEach to test run jest-axe before calling the cleanup function manually.

I'm not sure we want to be doing that for every test in the monorepo, do you know if there's a way to scope jest settings to an individual package? Maybe having each package author own their test script and using lerna?

**Comment by joshblack** [comment]
@vpicone I think we might be misunderstanding each other, it's not that we want it to run for every test scope as this wouldn't handle the `act` case where we want to run aXe after an interaction has occurred as this is a source of a11y violations that the team is running into currently.

We only want to test the following situations:

1. When `render` is called to check the initial markup
2. When `act` is called with a user interaction to make sure the markup ARIA is wired correctly (if necessary)

This 100% might not apply to every test btw, I just know we were chatting about this in the meeting yesterday and wanted to make sure we had a clear path forward with this.

**Comment by vpicone** [comment]
@joshblack Hmm yeah, it seems like we'd want to expectNoViolations every time the markup changes. I made the assumption that running after each test would capture/test the state of the dom after we tested the interactions. This would obviously result in a lot of duplicate axe tests since not all of the tests would have dom changes. 

It seems like for the most part, folks lean on testing the initial component render and if necessary calling it manually after interactions resulting in DOM changes. After playing with the role queries, I'm hopeful that our tests can be geared more towards proactive a11y testing rather than leaning to heavily on constantly running axe checks.

**Review by dakahn — APPROVED** [review]

**Comment by joshblack** [comment]
I think the types of mistakes that we'd want to make sure we address (as much as we can) are:

- Mistakes in ARIA due to lack of knowledge
- Mistakes in ARIA due to a mistake in the implementation

I think these can occur in the following events:

1. When the markup is initially rendered (validate HTML and default ARIA)
2. When an interaction is applied (validate interactive ARIA like `aria-expanded` or ARIA that references another node)
3. At the end of a sequence of interactions (validates interactive ARIA is reset or updated accordingly)

To ground this in an example, we could imagine writing a `tabs` component:

```html
<!-- (1) Validate default ARIA and HTML markup -->
<div class="tabs">
  <div role="tablist" aria-label="Tab label">
    <button role="tab" aria-selected="false" aria-controls="tabpanel-one" id="tab-one">
      <!-- ... -->
    </button>
    <button role="tab" aria-selected="false" aria-controls="tabpanel-two" id="tab-two">
      <!-- ... -->
    </button>
  </div>
  <div tabindex="0" role="tabpanel" id="tabpanel-one" aria-labelledby="tab-one" hidden="hidden">
    <!-- ... -->
  </div>
  <div tabindex="0" role="tabpanel" id="tabpanel-two" aria-labelledby="tab-two" hidden="hidden">
    <!-- ... -->
  </div>
</div>
```

```diff
<!-- (2) Validate an interaction where a tab is clicked -->
<div class="tabs">
  <div role="tablist" aria-label="Tab label">
-   <button role="tab" aria-selected="false" aria-controls="tabpanel-one" id="tab-one">
+   <button role="tab" aria-selected="true" aria-controls="tabpanel-one" id="tab-one">
      <!-- ... -->
    </button>
    <button role="tab" aria-selected="false" aria-controls="tabpanel-two" id="tab-two">
      <!-- ... -->
    </button>
  </div>
-  <div tabindex="0" role="tabpanel" id="tabpanel-one" aria-labelledby="tab-one" hidden="hidden">
+  <div tabindex="0" role="tabpanel" id="tabpanel-one" aria-labelledby="tab-one">

    <!-- ... -->
  </div>
  <div tabindex="0" role="tabpanel" id="tabpanel-two" aria-labelledby="tab-two" hidden="hidden">
    <!-- ... -->
  </div>
</div>
```

```diff
<!-- (3) Validate an interaction where a tab is clicked again -->
<!-- This one could 100% be unneeded, but we do need to make sure it happens somehow -->
<div class="tabs">
  <div role="tablist" aria-label="Tab label">
-   <button role="tab" aria-selected="true" aria-controls="tabpanel-one" id="tab-one">
+   <button role="tab" aria-selected="false" aria-controls="tabpanel-one" id="tab-one">
      <!-- ... -->
    </button>
    <button role="tab" aria-selected="false" aria-controls="tabpanel-two" id="tab-two">
      <!-- ... -->
    </button>
  </div>
-  <div tabindex="0" role="tabpanel" id="tabpanel-one" aria-labelledby="tab-one">
+  <div tabindex="0" role="tabpanel" id="tabpanel-one" aria-labelledby="tab-one" hidden="hidden">
    <!-- ... -->
  </div>
  <div tabindex="0" role="tabpanel" id="tabpanel-two" aria-labelledby="tab-two" hidden="hidden">
    <!-- ... -->
  </div>
</div>
```

This is to say that checking start/end and intermediate states can be helpful or necessary, but implementors may not be 100% certain of when to apply these strategies. This is why the idea of the tooling came up so that we could have a safety net to fall back on. Definitely down for really anything that can help with this, but just wanted to show a motivation for why this could be important and valuable for preventing accessibility regressions.

**Comment by vpicone** [comment]
@joshblack do you have an idea for how we could invoke axe during those scenarios aside from manually running calling it? Im not sure how we’d get jest to automatically infer where those places would be without over testing.

It seems to me that if we know the dom has changed, we can run axe manually, then eventually transition to explicitly testing for the correct roles. Since Axe estimates they catch only 30% of accessibility issues it might give a false sense of security around accurately testing these interactions forever.

**Comment by joshblack** [comment]
@vpicone this is where it was interesting to see if we could wrap RTL such that we could have the following:

```jsx
import { cleanup, render } from 'some-wrapper-module';

// Could handle (3), although maybe a better alternative
// is in the test
afterEach(cleanup);

test('some test', () => {
  // `render` wrapper handles situation (1)
  const { act, cleanup } = render(<SomeComponent />);

  // `act` wrapper handles situation (2)
  act(() => {
    // Interact with container in some way
  });

  // `cleanup` in this block could be another way to check
  cleanup();
});
```

Definitely agree with what you're saying, this could totally be an explicit check. Also 100% agreed on the 30% of issues point, I should have said earlier that really the only scope this stuff falls under is AVT1 with regards to structural verification. If we have keyboard helpers, we might be able to dip into AVT2 but for the most part AVT2/3 are out-of-scope 😞 

My understanding with the explicit checks is that we'd get tripped up by the first thing we'd want to protect against above, namely:

> Mistakes in ARIA due to lack of knowledge

In this case, we don't know what we should be testing. This is why I thought the addition of aXe here could be helpful (and hopefully later on DAP) until we scale out our knowledge. Definitely understand how futile it is though to believe that this would capture everything we don't know we don't know, but just was trying to bump our confidence here any way we can.

One downside of explicit checks could be that they may need to change as the rulesets themselves are updated, or made more strict, and I was thinking that relying on tooling to handle this would mean that we wouldn't have to stay on top of some of these updates. Instead, we could write our tests around the implementation and leverage tooling to handle these details instead of writing them out explicitly in our test suite. It also could mean writing the same kinds of tests across different test suites which always sucks as a test author.

Looking forward, understanding when DOM modification could occur could get gnarly when working in things like UI Shell or Data Table. I think this was why the idea of leveraging tooling was ultimately so interesting since we just write tests around usage and the tooling could capture AVT1 violations so we wouldn't have to think about this while writing tests, or duplicate these explicit checks across test files for certain types of patterns.

**Comment by vpicone** [comment]
@joshblack I think the implementation of a "smart" cleanup/render would be difficult, not sure how it'd infer that a dom change occurred since it's looking at essentially snapshots without context. Wrapping act/render would result in a lot of over-testing when an update doesn't change the dom. Even disabling the cleanup would result in duplicate tests of the same dom output.

At least initially I think the best approach would be to call explicitly after we anticipate dom updates. Once we get an idea for the class of errors this tool is helping us with, we might be better able to figure out when/where it can be automated in the testing lifecycle.

>just write tests around usage and the tooling could capture AVT1 violations so we wouldn't have to think about this while writing tests

I think this requires a careful blend of defensive/offensive approaches. I don't think we can afford to lean too heavily on the static analysis tooling being run constantly. Going of the types of errors this tool outputs, it seems mostly geared towards product code (image alt tags etc.) and catching the most critical role mismatches.

**Comment by joshblack** [comment]
What would you say the core issue is here @vpicone? Is it regarding how difficult an implementation might be or around the goals from above? Trying to figure out what to focus on here 👀 

---

> I think the implementation of a "smart" cleanup/render would be difficult, not sure how it'd infer that a dom change occurred since it's looking at essentially snapshots without context.

Most likely would not try and detect changes and would just run on the DOM node. We could compute a hash of the content and cache that such that we save results if performance is the concern here

> Wrapping act/render would result in a lot of over-testing when an update doesn't change the dom. Even disabling the cleanup would result in duplicate tests of the same dom output.

Seems like the cache solution above would help such that performance wouldn't be the issue. In other words, find a hash function for a given DOM node and sub-tree and cache the results of the first run.

> At least initially I think the best approach would be to call explicitly after we anticipate dom updates.

How would we know if a DOM update is supposed to be changing without prior knowledge? I think that's something that would be great if we could catch, assuming that we all know when something should be changing seems like we're setting ourselves up for issues that we don't know about that tooling like aXe and DAP help with.

>  I don't think we can afford to lean too heavily on the static analysis tooling being run constantly.

I think we're just using it for AVT1, AVT2/3 would still need to be done manually.

**Comment by vpicone** [comment]
@joshblack my concern is persisting state between tests (presumably a hash of the component) and false positives for DOM changes. I think it would restrict the way tests need to be written: providing wrong/slightly different dummy children in a subsequent test (even if it's just a change in string, using a different node) would result in triggering another check. This means we'd need to force consistency in children or end up silently running duplicate tests.

It would be a lot of complexity to introduce for a false sense of coverage on things that should be tested explicitly. Considering the number of interactive components with significant DOM changes on interaction, I think we're better off manually triggering a rerun.

I'd rather have a issues we miss at first, pick them up in manual testing (which will have to be a part of the process anyway) and reinforce with explicit tests than deal with the overhead of hashing/diffing tests outputs.

**Comment by vpicone** [comment]
Closing pursuant to more test refactor work

**Pull request closed without merging**